aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -73,177 +73,278 @@ void TensorIterator::reorder_dimensions() {
-}
-
-static std::tuple<Device, ScalarType, bool> compute_common_type_(at::ArrayRef<OperandInfo> operands) {
-  // See [Result type computation] in TensorIterator.h
-  auto device = compute_device(operands);
-  auto common_type = ScalarType::Undefined;
-  bool all_same_type = true;
-  for (const auto& op: operands){
-    if (!op.tensor.defined()) continue;
-    //don't handle scalars
-    if (op.tensor.dim() > 0){
-      ScalarType current = op.current_dtype;
-      if (current == ScalarType::Undefined){
-        all_same_type = false;
-        break;
-      }
-      if (common_type == ScalarType::Undefined) common_type = current;
-      if (common_type != current) {
-        all_same_type = false;
-        break;
-      }
-    } else {
-      all_same_type = false;
-      break;
-    }
-  }
-  if (all_same_type) {
-    return std::make_tuple(device, common_type, true);
+// See [Result type computation] in TensorIterator.h
+// NOTE: This is (or at least, was) slow, and compute_types() (below) attempts
+//   to avoid using it.
+ScalarType TensorIterator::compute_common_dtype() {
+  if (operands_.size() - num_outputs_ == 0) {
+    return ScalarType::Undefined;
   }
 
   at::native::ResultTypeState state = {};
-  for (const auto& op : operands) {
+  for (const auto& op : operands_) {
+    if (op.is_output) {
+      continue;
+    }
     state = at::native::update_result_type_state(op.tensor, state);
   }
-  auto dtype = at::native::result_type(state);
 
-  auto result = std::make_tuple(device, dtype, false);
-  TORCH_INTERNAL_ASSERT(dtype != ScalarType::Undefined);
-  return result;
-}
+  common_dtype_ = at::native::result_type(state);
+  TORCH_INTERNAL_ASSERT(common_dtype_ != ScalarType::Undefined);
+
+  return common_dtype_;
+}
+
+// compute_types() is misleadingly named since it does more than compute types.
+//   Its intent is to:
+//     (1) determine the operation's "common dtype" or, "computation dtype,"
+//           if necessary
+//     (2) check that all operands are on the same device, if requested
+//     (3) check that all operands share the same dtype, if requested
+//     (4) create temporaries as inputs and outputs, if needed and requested,
+//           on the CPU
+//   The CPU-specific behavior is because CUDA kernels are expected to
+//   cast to and from the computation dtype in the kernel itself.
+//
+//   Several flags control this behavior, they are:
+//     - check_all_same_dtype_, if true all operands must have the same dtype
+//     - check_all_same_device_, if true all operands must be on the same device
+//     - promote_inputs_to_common_dtype_, if true allows temporaries to be
+//         created for inputs on the CPU and sets input operands' target
+//         dtypes to the computation dtype
+//     - cast_common_dtype_to_outputs_, if true allows temporaries to be
+//         created to hold the results of the computation on the CPU and
+//         sets output operands' target dtypes to the computation dtype
```

Comment:
I don't know what this means. What's an operand "target dtype"? Isn't it more of a "source" dtype?

Reply:
That's a great question. I can probably add some additional commentary clarifying this.